src/evm/decoder.rs

```diff
@@ -0,0 +1,466 @@
+use crate::{
+    evm::{
+        engine_db::{update_engine, SHARED_TYCHO_DB},
+        tycho_models::{AccountUpdate, ResponseAccount},
+    },
+    models::Token,
+    protocol::{
+        errors::InvalidSnapshotError,
+        models::{BlockUpdate, ProtocolComponent, TryFromWithBlock},
+        state::ProtocolSim,
+    },
+};
+use alloy_primitives::Address;
+use std::{
+    collections::{hash_map::Entry, HashMap},
+    future::Future,
+    pin::Pin,
+    str::FromStr,
+    sync::Arc,
+};
+use thiserror::Error;
+use tokio::sync::RwLock;
+use tracing::{debug, info, warn};
+use tycho_client::feed::{synchronizer::ComponentWithState, FeedMessage, Header};
+use tycho_core::Bytes;
+
+#[derive(Error, Debug)]
+pub enum StreamDecodeError {
+    #[error("{0}")]
+    Fatal(String),
+}
+
+#[derive(Default)]
+struct DecoderState {
+    tokens: HashMap<Bytes, Token>,
+    states: HashMap<String, Box<dyn ProtocolSim>>,
+}
+
+type DecodeFut =
+    Pin<Box<dyn Future<Output = Result<Box<dyn ProtocolSim>, InvalidSnapshotError>> + Send + Sync>>;
+type RegistryFn =
+    dyn Fn(ComponentWithState, Header, Arc<RwLock<DecoderState>>) -> DecodeFut + Send + Sync;
+type FilterFn = fn(&ComponentWithState) -> bool;
+
+pub(super) struct TychoStreamDecoder {
+    state: Arc<RwLock<DecoderState>>,
+    skip_state_decode_failures: bool,
+    min_token_quality: u32,
+    registry: HashMap<String, Box<RegistryFn>>,
+    inclusion_filters: HashMap<String, FilterFn>,
+}
+
+impl TychoStreamDecoder {
+    pub fn new() -> Self {
+        Self {
+            state: Arc::new(RwLock::new(DecoderState::default())),
+            skip_state_decode_failures: false,
+            min_token_quality: 51,
+            registry: HashMap::new(),
+            inclusion_filters: HashMap::new(),
+        }
+    }
+
+    pub async fn set_tokens(&self, tokens: HashMap<Bytes, Token>) {
+        let mut guard = self.state.write().await;
+        guard.tokens = tokens;
+    }
+
+    // Method to register a decoder
```

Comment:
Pretty useless comment
```suggestion
```

Reply:
Yeah I need to still do the docs for this one